### PR TITLE
[Enhancement] Add a strict naming rule for graph names

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#957 

*Description of changes:*
The graph name must adhere to the Python identifier naming rules, with one exception that hyphens (-) are permitted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/source/cli/graph-construction/single-machine-gconstruct.rst [hunk 1]
```diff
@@ -33,7 +33,7 @@ Full argument list of the ``gconstruct.construct_graph`` command
 * **-\-num-processes-for-nodes**: the number of processes to process node data simultaneously. Increase this number can speed up node data processing.
 * **-\-num-processes-for-edges**: the number of processes to process edge data simultaneously. Increase this number can speed up edge data processing.
 * **-\-output-dir**: (**Required**) the path of the output data files.
-* **-\-graph-name**: (**Required**) the name assigned for the graph.
+* **-\-graph-name**: (**Required**) the name assigned for the graph. The graph name must adhere to the Python identifier naming rules with the exception that hyphens (``-``) are permitted.
```
Comment (zhjwy9343) [inline docs/source/cli/graph-construction/single-machine-gconstruct.rst:36]: Can we give a link of the `Python identifier naming rules` here, or provide some valid and invalid examples?

Comment (classicsong) [reply docs/source/cli/graph-construction/single-machine-gconstruct.rst]: Added

File: graphstorm-processing/graphstorm_processing/distributed_executor.py [hunk 2]
```diff
@@ -564,6 +569,33 @@ def parse_args() -> argparse.Namespace:
...
+    ---------
+    graph_name: str
+        Graph Name.
+    """
+    gname = re.sub(r"^\d+", "", graph_name)
+    assert gname.replace("-", "_").isidentifier(), (
+        "GraphStorm expects the graph name adheres to the Python"
+        "identifier naming rules with the exception that hyphens "
+        "(-) are permitted and the name can start with numbers. "
+        f"But we get {graph_name}"
```
Comment (thvasilo) [inline graphstorm-processing/graphstorm_processing/distributed_executor.py:595]:
```suggestion
        f"Got: {graph_name}"
```

File: python/graphstorm/utils.py [hunk 3]
```diff
@@ -31,6 +32,29 @@
...
+    ---------
+    graph_name: str
+        Graph Name.
+    """
+    gname = re.sub(r'^\d+', '', graph_name)
+    assert gname.replace('-', '_').isidentifier(), \
+        "GraphStorm expects the graph name adheres to the Python" \
+        "identifier naming rules with the exception that hyphens " \
+        "(-) are permitted and the name can start with numbers. " \
+        f"But we get {graph_name}."
```
Comment (thvasilo) [inline python/graphstorm/utils.py:56]:
```suggestion
        f"Got: {graph_name}."
```

File: tests/unit-tests/test_gsf.py [hunk 4]
```diff
@@ -449,7 +450,45 @@ def test_create_builtin_lp_decoder():
...
+        invalid_name = True
+    assert invalid_name
+
+    graph_name = "\graph_name"
+    invalid_name = False
+    try:
+        check_graph_name(graph_name)
+    except:
+        invalid_name = True
+    assert invalid_name
```
Comment (thvasilo) [inline tests/unit-tests/test_gsf.py:488]: Use `pytest.raises` to check for expected exceptions https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest.raises

**Review by thvasilo — CHANGES_REQUESTED** [review]

File: docs/source/cli/graph-construction/single-machine-gconstruct.rst [hunk 5]
```diff
@@ -33,7 +33,7 @@ Full argument list of the ``gconstruct.construct_graph`` command
 * **-\-num-processes-for-nodes**: the number of processes to process node data simultaneously. Increase this number can speed up node data processing.
 * **-\-num-processes-for-edges**: the number of processes to process edge data simultaneously. Increase this number can speed up edge data processing.
 * **-\-output-dir**: (**Required**) the path of the output data files.
-* **-\-graph-name**: (**Required**) the name assigned for the graph.
+* **-\-graph-name**: (**Required**) the name assigned for the graph. The graph name must adhere to the `Python identifier naming rules<https://docs.python.org/3/reference/lexical_analysis.html#identifiers>`_ with the exception that hyphens (``-``) are permitted and the name can start with numbers.
```
Comment (thvasilo) [inline docs/source/cli/graph-construction/single-machine-gconstruct.rst:36]:
```suggestion
* **-\-graph-name**: (**Required**) the name assigned for the graph. The graph name must adhere to the `Python identifier naming rules<https://docs.python.org/3/reference/lexical_analysis.html#identifiers>`_ with the exception that hyphens (``-``) are permitted and the name can start with numbers. e.g. "2024-graph" and "my_graph" are valid names, but "my.graph" and "my/graph" are not.
```

File: docs/source/cli/model-training-inference/distributed/sagemaker.rst [hunk 6]
```diff
@@ -388,7 +388,7 @@ The rest of the arguments are passed on to ``sagemaker_train.py`` or ``sagemaker
 
 * **--task-type**: Task type.
 * **--graph-data-s3**: S3 location of the input graph.
-* **--graph-name**: Name of the input graph.
+* **--graph-name**: Name of the input graph. The graph name must adhere to the `Python identifier naming rules<https://docs.python.org/3/reference/lexical_analysis.html#identifiers>`_ with the exception that hyphens (``-``) are permitted and the name can start with numbers.
```
Comment (thvasilo) [inline docs/source/cli/model-training-inference/distributed/sagemaker.rst:391]:
```suggestion
* **--graph-name**: Name of the input graph. The graph name must adhere to the `Python identifier naming rules<https://docs.python.org/3/reference/lexical_analysis.html#identifiers>`_ with the exception that hyphens (``-``) are permitted and the name can start with numbers. e.g. "2024-graph" and "my_graph" are valid names, but "my.graph" and "my/graph" are not.
```

**Review by thvasilo — APPROVED** [review]
LGTM now, small comments to add a couple of examples.

File: docs/source/cli/graph-construction/single-machine-gconstruct.rst (hunk 5 above)
Comment (classicsong) [reply docs/source/cli/graph-construction/single-machine-gconstruct.rst:36]: I do not think we need to give examples. There are too much details.

**Pull request merged**